src/infra/core/src/transform_service_impl.rs

```diff
@@ -333,16 +312,29 @@ impl TransformServiceImpl {
     // TODO: Allow derivative datasets to function with inputs containing no data
     // This will require passing the schema explicitly instead of relying on a file
     async fn is_never_pulled(&self, dataset_ref: &DatasetRef) -> Result<bool, InternalError> {
-        let dataset = self.dataset_repo.get_dataset(dataset_ref).await.int_err()?;
-        let last_data_block = dataset
-            .as_metadata_chain()
-            .iter_blocks()
-            .filter_data_stream_blocks()
-            .try_next()
+        type Flag = MetadataEventTypeFlags;
+        type Decision = MetadataVisitorDecision;
+
+        Ok(self
+            .dataset_repo
+            .get_dataset(dataset_ref)
             .await
-            .int_err()?;
-        Ok(last_data_block
-            .and_then(|(_, b)| b.event.last_offset())
+            .int_err()?
+            .as_metadata_chain()
+            .reduce(
+                None,
+                Decision::NextOfType(Flag::DATA_BLOCK),
+                |state, _, block| {
+                    let Some(data_block) = block.as_data_stream_block() else {
```

Comment:
Given that data blocks are 99.9% of dataset the old version seemed simpler and as effective.

Reply:
I think you might like the new version better -- Done